src/cli.rs

```diff
@@ -0,0 +1,153 @@
+use crate::{merkle_tree::MerkleTree, proof_element::ProofElement};
+use anyhow::{Error, Result};
+use clap::{Parser, Subcommand};
+use std::{
+    fs::File,
+    io::{self, BufRead, BufReader, Write},
+    path::PathBuf,
+};
+
+/// CLI tool for tree-related operations
+#[derive(Parser, Debug)]
+#[command(name = "tree")]
+#[command(about = "Tree command-line tool", long_about = None)]
+pub struct Cli {
+    #[command(subcommand)]
+    command: Commands,
+
+    /// Merkle tree instance to hold the state
+    #[arg(skip)]
+    mktree: MerkleTree,
+}
+
+#[derive(Subcommand, Debug)]
+enum Commands {
+    /// Shows the tree structure
+    Show,
+    /// Adds an element to the tree
+    Add { element: String },
+    /// Verifies a proof for a given hash
+    Verify { hash: String, proof_file: PathBuf },
+    /// Generates a proof for a given hash.
+    Proof { hash: String },
+    /// Builds a tree with the provided elements.
+    Build { elements: Vec<String> },
+}
+
+impl Default for Cli {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Cli {
+    pub fn new() -> Self {
+        Cli {
+            command: Commands::Show, // Placeholder, will be overridden by parsing
+            mktree: MerkleTree::new_empty(),
+        }
+    }
+
+    pub fn run(&mut self) -> Result<()> {
+        loop {
+            // Display prompt
+            print!("tree> ");
+            io::stdout().flush()?; // Flush prompt to the terminal
```

Comment:
For example here what I would do is the following:

```suggestion
            io::stdout().flush().map_err(|e| MerkleError::FailedToFlush(e.to_string()))?; // Flush prompt to the terminal
```